worker/caasoperator/caasoperator.go

```diff
@@ -452,45 +467,63 @@ func (op *caasOperator) loop() (err error) {
 					return errors.Trace(err)
 				}
 				// Notify all uniters of the change so they run the upgrade-charm hook.
-				for unitId, changedChan := range aliveUnits {
-					logger.Debugf("trigger upgrade charm for caas unit %v", unitId)
+				for unitID, changedChan := range aliveUnits {
+					logger.Debugf("trigger upgrade charm for caas unit %v", unitID)
 					select {
 					case <-op.catacomb.Dying():
 						return op.catacomb.ErrDying()
 					case changedChan <- struct{}{}:
 					}
 				}
 			}
+		case units, ok := <-containerStartWatcher.Changes():
+			if !ok {
+				return errors.New("container start watcher closed channel")
+			}
+			for _, unitID := range units {
+				if runningChan, ok := unitRunningChannels[unitID]; ok {
+					logger.Debugf("trigger running status for caas unit %v", unitID)
+					select {
+					case <-op.catacomb.Dying():
+						return op.catacomb.ErrDying()
+					case runningChan <- struct{}{}:
```

Comment:
We should consider the case where the unit goes to dead (below). We should abort any attempt to send on the runningChan. So we should create a unitDead chan or something and close it below and select on it here

Reply:
We have the same problem with the aliveUnits chans, I don't think it's a terrible problem since the uniter worker restarts and uses the same channels.

Proper way to deal with this is have the resolver create watchers for both application changes and running status changes, that way the watcher is tied to the lifecycle of the resolver/uniter.
Then we can use the watcher's dying channel to safely notify.